nexus/src/app/background/tasks/instance_watcher.rs

```diff
@@ -90,96 +93,123 @@ impl InstanceWatcher {
-                .vmm_get_state(&vmm_id)
-                .await
-                .map_err(SledAgentInstanceError);
             let mut check = Check {
                 target,
                 outcome: Default::default(),
                 result: Ok(()),
                 update_saga_queued: false,
             };
-            let state: SledVmmState = match rsp {
-                Ok(rsp) => rsp.into_inner().into(),
-                // Oh, this error indicates that the VMM should transition to
-                // `Failed`. Let's synthesize a `SledInstanceState` that does
-                // that.
-                Err(e) if e.vmm_gone() => {
-                    slog::info!(
-                        opctx.log,
-                        "sled-agent error indicates that this instance's \
-                         VMM has failed!";
-                        "error" => %e,
-                    );
-                    check.outcome =
-                        CheckOutcome::Failure(Failure::NoSuchInstance);
-                    // TODO(eliza): it would be nicer if this used the same
-                    // code path as `mark_instance_failed`...
-                    SledVmmState {
-                        vmm_state: nexus::VmmRuntimeState {
-                            r#gen: vmm.runtime.r#gen.0.next(),
-                            state: nexus::VmmState::Failed,
-                            time_updated: chrono::Utc::now(),
-                        },
-                        // It's fine to synthesize `None`s here because a `None`
-                        // just means "don't update the migration state", not
-                        // "there is no migration".
-                        migration_in: None,
-                        migration_out: None,
-                    }
+
+            let state = if sled.policy() == SledPolicy::Expunged {
+                // If the sled has been expunged, any VMMs still on that sled
+                // should be marked as `Failed`.
+                slog::info!(
+                    opctx.log,
+                    "instance is assigned to a VMM on an Expunged sled; \
+                     marking it as Failed";
+                );
+                check.outcome = CheckOutcome::Failure(Failure::SledExpunged);
+                // TODO(eliza): it would be nicer if this used the same
+                // code path as `mark_instance_failed`...
+                SledVmmState {
+                    vmm_state: nexus::VmmRuntimeState {
+                        r#gen: vmm.runtime.r#gen.0.next(),
+                        state: nexus::VmmState::Failed,
+                        time_updated: chrono::Utc::now(),
+                    },
+                    // It's fine to synthesize `None`s here because a `None`
+                    // just means "don't update the migration state", not
+                    // "there is no migration".
+                    migration_in: None,
+                    migration_out: None,
```

Comment:
Just to refresh my memory: if there *was* an in-doubt migration, it will get marked as failed when we next reconcile the instance and VMM records in the update saga, correct?

Reply:
Hmm, it looks like we don't currently do that --- if an instance that has an active migration sees its active VMM go to `Failed` in a `SledVmmState` update that doesn't include a migration update, the migration will remain in its prior state. We should probably fix that, and make sure that any migration involving a `Failed` VMM also goes to `Failed`. If you don't mind, I'd prefer to do that in a separate branch, as it's not _directly_ related to this change in particular?

I think you're right that the update saga should probably be responsible for failing active migrations when a VMM goes to `Failed`, because doing it in `process_instance_update` is a bit annoying: we would want to fail a migration if there is one, but not accidentally _create_ a new migration just to fail it.

Reply:
A separate branch is fine for this, thanks! I took the liberty of filing #6520 to track.

Reply:
Great, thank you!